lib/puppet/provider/user/useradd.rb

```diff
@@ -152,6 +152,37 @@ def groups=(value)
     set(:groups, value)
   end
 
+  def password=(value)
+    user = @resource[:name]
+    tempfile = Tempfile.new('puppet', :encoding => Encoding::ASCII)
+    begin
+      # Puppet execute does not support strings as input, only files.
+      # The password is expected to be in an encrypted format given -e is specified:
+      tempfile << "#{user}:#{value}\n"
```

Comment:
Can `user` contain non-ASCII like `Gérard`? If so then I think we need to open the tempfile using `:encoding => Encoding.default_external`, which will typically be `UTF-8`

Reply:
It seems that users can contain non-ASCII characters (like `Gérard`), on linux, but Encoding.default_external on Debian 8 and 10 returned 

```sh-session
➜ facter os.distro.description
Debian GNU/Linux 10 (buster)
➜ echo $LC_CTYPE
UTF-8
➜ irb
irb(main):001:0> Encoding.default_external
=> #<Encoding:US-ASCII>
```

Using `Encoding.default_external` in this case resulted in

```sh-session
➜ puppet apply manifest.pp
    157:     tempfile = Tempfile.new('puppet', :encoding => Encoding::default_external)
    158:     begin
    159:       # Puppet execute does not support strings as input, only files.
    160:       # The password is expected to be in an encrypted format given -e is specified:
    161:       binding.pry
 => 162:       tempfile << "#{user}:#{value}\n"
    163:       tempfile.flush
    164:
    165:       # Options '-e' use encrypted password
    166:       # Must receive "user:enc_password" as input
    167:       # command, arguments = {:failonfail => true, :combine => true}

➜ [1] pry(#<Puppet::Type::User::ProviderUseradd>)> user
=> "G\u00E9rard"
➜ [2] pry(#<Puppet::Type::User::ProviderUseradd>)> user.encoding
=> #<Encoding:UTF-8>
➜ [3] pry(#<Puppet::Type::User::ProviderUseradd>)> value
=> "pass"
➜ [4] pry(#<Puppet::Type::User::ProviderUseradd>)> value.encoding
=> #<Encoding:UTF-8>
➜ [5] pry(#<Puppet::Type::User::ProviderUseradd>)> Encoding::default_external
=> #<Encoding:US-ASCII>
➜ [6] pry(#<Puppet::Type::User::ProviderUseradd>)> continue
Error: Could not set password on user[Gérard]: U+00E9 from UTF-8 to US-ASCII
Error: /Stage[main]/Main/User[Gérard]/password: change from [redacted] to [redacted] failed: Could not set password on user[Gérard]: U+00E9 from UTF-8 to US-ASCII
Notice: Applied catalog in 21.13 seconds
```
so I hardcoded it to `Encoding::UTF_8`. 
@joshcooper, any concerns regarding this?

Reply:
Also while testing this, we found out another issue when querying users with non-ASCII characters.
Ticket created for this: https://tickets.puppetlabs.com/browse/PUP-11137

Reply:
Ok I think I see what's going on. First, I was using a VM from the pooler and ruby's default external encoding is UTF8:

```
root@drier-caramel:~# facter os.distro.description
Debian GNU/Linux 10 (buster)
root@drier-caramel:~# locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
root@drier-caramel:~# ruby -e "puts Encoding.default_external"
UTF-8
```

At first I wasn't able to reproduce the error using `Encoding.default_external`. Then I tried changing my system locale to `de_DE.iso88591`

```shell
# dpkg-reconfigure locales
...
# locale
LANG=de_DE.iso88591
LANGUAGE=
LC_CTYPE="de_DE.iso88591"
LC_NUMERIC="de_DE.iso88591"
LC_TIME="de_DE.iso88591"
LC_COLLATE="de_DE.iso88591"
LC_MONETARY="de_DE.iso88591"
LC_MESSAGES="de_DE.iso88591"
LC_PAPER="de_DE.iso88591"
LC_NAME="de_DE.iso88591"
LC_ADDRESS="de_DE.iso88591"
LC_TELEPHONE="de_DE.iso88591"
LC_MEASUREMENT="de_DE.iso88591"
LC_IDENTIFICATION="de_DE.iso88591"
LC_ALL=de_DE.iso88591
```

Then I applied your patch and switched it to using `Encoding.default_encoding` and I got the same error:

```
Error: /Stage[main]/Main/User[Gérard]/password: change from [redacted] to [redacted] failed: incompatible character encodings: UTF-8 and ISO-8859-1
```

It took me awhile to understand the exception. The problem is that `chpasswd` is unable to find the user account that we specified (more on that later). So `chpasswd` returns non-empty output and we [raise an `ExecutionFailure` with the output in the message](https://github.com/puppetlabs/puppet/pull/8661/files#diff-2ebf83c6948b2c0119d093d483d0637fa57da020174d28af4a007a875d92987eR178).

We rescue the exception and try to [interpolate a string containing the resource's name, encoded as `UTF-8`, and the `detail` exception message, encoded as `ISO-8859-1`](https://github.com/puppetlabs/puppet/pull/8661/files#diff-2ebf83c6948b2c0119d093d483d0637fa57da020174d28af4a007a875d92987eR182).

```
(byebug) Encoding.default_external
#<Encoding:ISO-8859-1>
(byebug) @resource.name
"G\u00E9rard"
(byebug) @resource.name.encoding
#<Encoding:UTF-8>
(byebug) detail.message.encoding
#<Encoding:ISO-8859-1>
(byebug) "#{@resource.name} #{detail}"
*** Encoding::CompatibilityError Exception: incompatible character encodings: UTF-8 and ISO-8859-1
```

So why did `chpasswd` fail? The reason is if we write the file using `Encoding.default_external`, then on disk we'll have:

```
0000000 47 e9 72 61 72 64 3a 47 e9 72 61 72 64 31 36 32  >G�rard:G�rard162<
0000020 35 31 30 33 30 30 31 0a                          >5103001.<
```

Note `é` is defined as `\xE9` in https://en.wikipedia.org/wiki/ISO/IEC_8859-1

However, after forking the child process and before executing `chpasswd` we [switch to the C locale](https://github.com/puppetlabs/puppet/blob/acefb78141b5b647cf10c2b32fe901b201d941b9/lib/puppet/util/execution.rb#L344-L348). My understanding is that it forces `chpasswd` to use the [Portable Character Set](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap06.html) and `\xE9` doesn't appear to be in that set. So then `chpasswd` can't find the user and emits:

```
(byebug) puts output.encoding
ISO-8859-1
nil
(byebug) puts output.encode(Encoding::UTF_8)
chpasswd: line 1: user 'Gérard' does not exist
chpasswd: error detected, changes ignored
```

So I think the correct behavior is to encode the tempfile content in a way that the `chpasswd` running in the C locale can read it without loss of information. Using `ASCII` doesn't work, but `UTF-8` does. I'm not entirely sure why `UTF-8` works though, because the `é` is encoded as `\xC3 \xA9` and those aren't in the portable set either.

Reply:
I ran an ad-hoc with tests with password as `Gérard` to see if there is some differences before the fix when displaying the user, and I think it is ok. Even if password is not displayed as `Gérard` they both work the same.

On 6.9.0:
```
[root@wily-armful ~]# puppet resource user ggg ensure=present provider=useradd password=Gérard
2021-07-05 14:45:31.853026 WARN  puppetlabs.facter - locale environment variables were bad; continuing with LANG=C LC_ALL=C
Notice: /User[ggg]/ensure: created
user { 'ggg':
  ensure   => 'present',
  password => 'GÃ©rard',
}
[root@wily-armful ~]# puppet resource user ggg
2021-07-05 14:45:46.379401 WARN  puppetlabs.facter - locale environment variables were bad; continuing with LANG=C LC_ALL=C
user { 'ggg':
  ensure             => 'present',
  gid                => 1000,
  home               => '/home/ggg',
  password           => 'GÃ©rard',
  password_max_age   => 99999,
  password_min_age   => 0,
  password_warn_days => 7,
  shell              => '/bin/bash',
  uid                => 1000,
}
```
Is there anything I can do here, or is encoding the tempfile as UTF-8 enough `tempfile = Tempfile.new('puppet', :encoding => Encoding::UTF_8)`